[dtensor] early return for _split_tensor

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #125810

as titled, if _split_tensor does not require padding or even is evenly
sharded on the dim, no need to calculate padding and could simply return

This is to avoid some unnecessary CPU operations

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

wanchaol committed:
[dtensor] early return for _split_tensor

as titled, if _split_tensor does not require padding or even is evenly
sharded on the dim, no need to calculate padding and could simply return

This is to avoid some unnecessary CPU operations

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] early return for _split_tensor"


as titled, if _split_tensor does not require padding or even is evenly
sharded on the dim, no need to calculate padding and could simply return

This is to avoid some unnecessary CPU operations

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

wz337 reviewed (approved):
This is great! LGTM!

XilunWu commented on torch/distributed/_tensor/placement_types.py:
```diff
@@ -60,9 +60,17 @@ def _split_tensor(
+        if not with_padding or tensor.size(self.dim) % num_chunks == 0:
+            if contiguous:
+                tensor_list = [t.contiguous() for t in tensor_list]
+            return tensor_list, []
```
directly returning seems problematic according to the test:
```
l = torch.chunk(torch.rand(5, 2), 4)  # yields 3 smaller tensors of size (2, 2), (2, 2), (1, 2)
l[3]  # index out of range
```

wz337 replied:
In this case, with_padding would be True and tensor.size(self.dim) % num_chunks won't be 0, right? So it shouldn't go in this if, if I am understanding this correctly.

XilunWu replied:
The failed test case has `with_padding=False`.

wanchaol replied:
fixed the tests

XilunWu reviewed (dismissed):
Need fix test.

wanchaol committed:
Update on "[dtensor] early return for _split_tensor"


as titled, if _split_tensor does not require padding or even is evenly
sharded on the dim, no need to calculate padding and could simply return

This is to avoid some unnecessary CPU operations

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] early return for _split_tensor"


as titled, if _split_tensor does not require padding or even is evenly
sharded on the dim, no need to calculate padding and could simply return

This is to avoid some unnecessary CPU operations

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] early return for _split_tensor"


as titled, if _split_tensor does not require padding or even is evenly
sharded on the dim, no need to calculate padding and could simply return

This is to avoid some unnecessary CPU operations

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]
